testing/ef-tests/src/cases/blockchain_test.rs

```diff
@@ -117,80 +185,147 @@ impl Case for BlockchainTestCase {
+                OriginalValuesKnown::Yes,
+                StorageLocation::Database,
+            )
+            .map_err(|_| Error::BlockProcessingFailed { block_number })?;
+
+        provider
+            .write_hashed_state(&hashed_state.into_sorted())
+            .map_err(|_| Error::BlockProcessingFailed { block_number })?;
+        provider
+            .update_history_indices(block.number..=block.number)
+            .map_err(|_| Error::BlockProcessingFailed { block_number })?;
+
+        // Since there were no errors, update the parent block
+        parent = block.clone()
+    }
+
+    // Validate the post-state for the test case.
+    //
+    // If we get here then it means that the post-state root checks
+    // made after we execute each block was successful.
+    //
+    // If an error occurs here, then it is:
+    // - Either an issue with the test setup
+    // - Possibly an error in the test case where the post-state root in the last block does not
+    //   match the post-state values.
+    let expected_post_state = case.post_state.as_ref().ok_or(Error::MissingPostState)?;
+    for (&address, account) in expected_post_state {
+        account.assert_db(address, provider.tx_ref())?;
+    }
+
+    Ok(())
+}
+
+fn decode_blocks(
+    test_case_blocks: &[crate::models::Block],
+) -> Result<Vec<RecoveredBlock<Block>>, Error> {
+    let mut blocks = Vec::with_capacity(test_case_blocks.len());
+    for (block_index, block) in test_case_blocks.iter().enumerate() {
+        // The blocks do not include the genesis block which is why we have the plus one.
+        // We also cannot use block.number because for invalid blocks, this may be incorrect.
+        let block_number = (block_index + 1) as u64;
+
+        let decoded = SealedBlock::<Block>::decode(&mut block.rlp.as_ref())
+            .map_err(|_| Error::BlockProcessingFailed { block_number })?;
+
+        let recovered_block = decoded
+            .clone()
+            .try_recover()
+            .map_err(|_| Error::BlockProcessingFailed { block_number })?;
+
+        blocks.push(recovered_block);
     }
 
-    // Drop the provider without committing to the database.
-    drop(provider);
+    Ok(blocks)
+}
+
+fn pre_execution_checks(
+    chain_spec: Arc<ChainSpec>,
+    parent: &RecoveredBlock<Block>,
+    block: &RecoveredBlock<Block>,
+) -> Result<(), Error> {
+    let consensus: EthBeaconConsensus<ChainSpec> = EthBeaconConsensus::new(chain_spec);
+
+    let sealed_header = block.sealed_header();
+    let header = block.header();
+
+    <EthBeaconConsensus<ChainSpec> as Consensus<Block>>::validate_body_against_header(
+        &consensus,
+        block.body(),
+        sealed_header,
+    )?;
+    consensus.validate_header_against_parent(sealed_header, parent.sealed_header())?;
+    consensus.validate_header_with_total_difficulty(header, block.difficulty)?;
+    consensus.validate_header(sealed_header)?;
+    consensus.validate_block_pre_execution(block)?;
```

Comment:
@kevaundray do we care about order of operations here? should it resemble the spec or how reth's tree engine performs consensus checks?

Reply:
For the tests, I think it doesn't matter because it doesn't check for the exact reason it failed. Happy to change it to match the engine tree

Reply:
I think this is fine, doesn't hurt to check this here as well